### ISSUE TITLE
Add editorial / critique markup

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#1031]
We should consider adding critique markup using one of the following two approaches:
- Leverage roles on the unquoted formatting mark to relate critique information (e.g., `[+]#inserted text#`)
- Integrate with critic markup (see http://criticmarkup.com/)

If we decide to implement the second option, it should be done as an alternative extension. Perhaps the best way to start is to prototype it in the extensions lab.

**Comment by CodeGnome** [comment]
Is change-tracking really the responsibility of a format processor? I respectfully suggest that change-tracking and displaying editorial markup isn't within the proper scope of Asciidoctor. Typical display markup such as [line-through][3] is already supported by AsciiDoc, and editorial insertions can be displayed as inline text like `Hi, [she] said.` Editorial markup can also be placed in [comment blocks][1] if the critique should be visible in the source document, but omitted from the rendered output.

Line-wise, word-wise, and character-wise diffs are best left to purpose-built tools. See [this post][2]  for a representative  cross-section of such tools.

In addition, multi-user editorial markup creates the problems of *attribution* and *precedence*. In other words, did Bob or Alice make the change? If both made changes, who made the change first, and whose change should take precedence? A modern SCM like Git already handles issues like these quite well, so I'm not sure I understand the use case for duplicating the functionality within Asciidoctor.

Another thing to consider is that editorial markup is rarely appropriate for display purposes. Marking up the markup is useful for communicating about changes that need to be made to the source document, but the practical use cases for *displaying* the editorial marks in an output document (as opposed to the source document) would seem rare.

I think there's a great deal of usefulness in allowing for editorial markup in source documents, but very little for such markup in rendered output documents. That's just my $0.02, of course, but I think it's a viewpoint worth considering.
 
[1]: http://asciidoc.org/userguide.html#X26
[2]: http://discuss.asciidoctor.org/Asciidoc-diffs-tp2533p2702.html
[3]: http://asciidoc.org/userguide.html#X51

**Comment by jxxcarlson** [comment]
I would like to lobby for a simple, quickly-typable form 
of markup in Asciidoctor for line-through, which
has very wide utility and for which the lack has
felt to me lik a missing limb.

Both  -foo- and =foo= would work — they don’t seem
to be taken.  I would favor =foo= since it is more visible
in the source text.

The 

[blue line-through]#foo bar#

option works but it is entirely too cumbersome. For that 
reason I doubt that humans use it much.

(I like the blue color, but this cold be handled in CSS).

Please say yes to this one, or something close to it!

@jxxcarlson

http://noteshare.io

**Comment by jirutka** [comment]
I would prefer `~~foo~~`, this syntax uses GitHub Flavored Markdown and Pandoc Markdown.

> Both -foo- and =foo= would work…

I’m afraid that these can be simply confused with regular text.

**Comment by jxxcarlson** [comment]
Despite being a bit awkward to type, this (~~foo~~) is what we should do.

I vote for it!

**Comment by jxxcarlson** [comment]
Regarding the appropriateness of editorial marks for Asciidoctor, 
imagine this scenario:

An English teacher (substitute your language for English of course)
wishes to post writing samples on her site so that students
can learn by example.  She has a great collection of editorially marked 
up essays, and needs a way of displaying the marks when she posts them …

… happily, the designers of Asciidoctor have already thought this through,
and it is not only doable, but doable in a convenient way.

(Yay!)

**Comment by mojavelinux** [comment]
On Sat, Jan 24, 2015 at 9:00 PM, Todd A. Jacobs <notifications@github.com>
wrote:

> Is change-tracking really the responsibility of a format processor? I
> respectfully suggest that change-tracking and displaying editorial markup
> isn't within the proper scope of Asciidoctor.



My instinct always leads me to the same conclusion. Since we have some
editorial markup in AsciiDoc already, it doesn't hurt to complete a basic
set by adding dedicated markup for strike-through at least.

A thorough critic markup integration should be done as an extension. It can
live in the Asciidoctor ecosystem, but it wouldn't live in the core
processor.

-Dan


-- 
Dan Allen | http://google.com/profiles/dan.j.allen

**Comment by serussell** [comment]
I'm eager for this, so I'd like to bump and ask if there's been any further thinking / progress on this front?  Has it gained, or lost, appeal to the heavy lifters (Dan, Ryan) in the past ten months?

**Comment by elextr** [comment]
The question of adding critic markup was asked on the Asciidoc list as well.

To summarise my thoughts, VCS systems do a better job of change tracking in a distributed development environment than using markup that makes the input file more messy as changes accumulate.  Critic markup seems more relevant to an environment where a single set of changes are made once and the document published, than when documents evolve as the associated systems do.

To repeat the conclusion: "So although its probably implementable, it doesn't seem to solve a problem thats relevant to todays development environment."

Full reply is [here](https://groups.google.com/d/msg/asciidoc/sAfj3Lf8jd0/xeSSL15ZDAAJ).

**Comment by obilodeau** [comment]
I agree with @elextr on that one. Here's the workflow we use for changes/reviews at work (small team of people familiar with asciidoc/git):

```
== Fixes

Typos and obviously bad wording should be fixed inplace. One can word-diff with
git to review what someone else fixed even if there would be white space 
issues.

    git diff --word-diff

== Comments or suggestions

To leverage the asciidoc markup language do your comments with lines starting
with //.

For example:

...
operation are usually able to remove other malware components such as
Linux/Cdorked, but often overlook the OpenSSH backdoor due to the stealthiness
//                                                                ^ sure about this word?
//                                                                I don't like it [OB]
mechanisms used. Thus, it was common for the operators to come back a few days
later and revert the changes made by the administrator.
// [OB] thinking about it, why don't you remove this whole paragraph?

Longer comments can be made using the block comment markup: ////

== Review branches

Reviews can be done inside branches where all suggestions are committed and
then the text reworded by the integrator as he accepts or rejects changes and
modify text according to comments. Then a squash merge of the review branch
can be done in master branch in order to keep the history tidy.
```

With this, we never felt the need for something like CriticMarkup.

Full asciidoc mailing list reply is [here](https://groups.google.com/d/msg/asciidoc/sAfj3Lf8jd0/Aj5hM4a3DAAJ).

**Comment by zaxebo1** [comment]
@obilodeau : git is very useful, but it does NOT cover every "use case". 

git policy can "not" be enforced on everybody, when you have to collaborate outside your organization walls. Most of the academicians collaborate through emails/email-attachment.

====

We collborate with other academicians of various other organizations, and communication happen only through email. I am trying to switch from MSOffice document(and sometimes latex documents) To asciidoctor based documents in this case, and everybody has agreed. Now if we force them to use git (instead of email), then nobody shall agree.
Mostly people use to "record the changes in msoffice during critique/editing the original document", In this case adding critique markup to asciidoctor is perfect usecase, for emulating the same in asciidoctor.

**Comment by jxxcarlson** [comment]
I’d just like to remark that Asciidoctor’s uses go well beyond development environments.
See,for example, http://www.noteshare.io <http://www.noteshare.io/>.

**Comment by elextr** [comment]
On 1 November 2015 at 16:43, zaxebo1 <notifications@github.com> wrote:

> @obilodeau <https://github.com/obilodeau> : git is very useful, but it
> does NOT cover every "use case".
>
> git policy can "not" be enforced on everybody, when you have to
> collaborate outside your organization walls. Most of the academicians
> collaborate through emails/email-attachment.
>
> We collborate with other academicians of various other organizations, and
> communication happen only through email. I am trying to switch from
> MSOffice document(and sometimes latex documents) To asciidoctor based
> documents in this case, and everybody has agreed. Now if we force them to
> use git (instead of email), then nobody shall agree.
> Mostly people use to "record the changes in msoffice during
> critique/editing the original document", In this case adding critique
> markup to asciidoctor is perfect usecase, for emulating the same in
> asciidoctor.
>

Of course you have a specific use-case and as Dan said you can support it
via an extension, it doesn't have to be in core.​  And if a sufficient
audience for it becomes apparent I'm sure the extension can move to being
part of the asciidoctor project.

And you can use the markup in comments to get some experience with it until
you get the extension implemented.

I will be interested in how you merge the comments from multiple
reviewers.  That seems to me to need to be manual with this workflow, have
you any ideas?

Cheers
Lex

**Comment by nicorikken** [comment]
What options might suit a note best for the final rendering? I tried a sidebar block and could see the use-case for a attribute block. Both however fail to link to a specific position in the scentence. An Alternative would be to have two styles: a block note, and an inline note. Regarding scentence position using some references in the text could be an option: like `[N12]` for note 12.

**Comment by obilodeau** [comment]
If you want a note pinned to a specific location, what you are looking for
is a footnote. There is a macro for that in asciidoc and I believe
asciidoctor too.

**Comment by nicorikken** [comment]
@obilodeau I should have thought about that. But looking into the options, It seems the footnotes will always be rendered at the bottom of the page (which could be very long).
As the notes are a one-off, and are only relevant to the direct context (say the paragraph). Having the option to render footnotes at the end of the paragraph rather than at the end of the page would be a nice alternative.
In general though, using a similar syntax for placing notes would be nice, although the placement could be slightly more local. Still the same styles could still apply.

**Comment by danyill** [comment]
I love Asciidoctor, I love Git. 

But I don't think these substitute for a way to create in the document (and also in rendered output) review comments, suggestions, changes etc. Though I love very little about Microsoft Word, I do like the reviewing pane bubbles and at least the pre-2016 the way it can readily show changes between versions with the author, date and time and the use of different colours.

Sure I can do a word diff in Git, but I feel the lack a tool which allows me to bring together comments from reviewers visually or see areas of change or concern in a document. Particularly for non-technical users who don't want to comb the git history, they just want a single summary document to look over (and probably print out), probably in rendered output (so they aren't required to understand Asciidoc). That's an end-user/reviewer as opposed to writer/developer function. The writer/developer, however has similar needs but in the source file.

I expect an extension which maybe also acts on the git revision history to produce a new Asciidoc file would be quite a neat idea. From what I read it needs to be in the AST and not just a markup for it to be "well-defined" and not conflict with other syntax.

**Comment by elextr** [comment]
> Sure I can do a word diff in Git, but I feel the lack a tool which allows me to bring together comments from reviewers visually or see areas of change or concern in a document.

I don't think Git was proposed as a review tool, its a change recorder.  Git can only record the changes that have been made.  With HUMAN help it can allow you to merge comments from multiple people, but it only sees differences between the repository and one set of changes that are proposed but not yet committed.  It can't compare multiple sets of changes from different people.  Once you have decided what changes to accept then Git will record them faithfully and tell you the history of who did what.

Combining multiple reviews is a separate process that needs a separate tool which is neither Git nor Asciidoctor, something perhaps like the MS Word capability that can visualise and combine multiple reviews.  Such a tool may generate an Asciidoc markup showing all the suggestions, but that needs no new markup in Asciidoc itself.

That tool is something outside the scope of Asciidoctor itself.

**Comment by ilvetz** [comment]
Yesterday my boss asked me for something and this function could give him a satisfactory answer. He asked me to highlight only new parts of a document I'm writing which he should review and approve.

He knows about the CHANGELOG section which I always update when submitting but, since the document is very long, he finds uncomfortable to scroll the entire document looking for new paragraphs.

I find really useful to show corrections and/or to highlight updates or selected sections of the document.

**Comment by maxandersen** [comment]
> doc.find_by(context: :paragraph).each {|block|
>   source = block.source
>   # process criticmarkup on source, replace with valid inline AsciiDoc
>   source.lines.replace block.source.split("\n")
> }

so I got the basics working, but I can for the love of me not figure out how to actually replace content in the treeporcessor.

in the example above you get an error that "replace" method does not exist.

Can someone help on how to actually adjust the text in the treeprocessor ?

**Comment by serussell** [comment]
@maxandersen, have you tried:
```
    source.lines = block.source.split("\n")
```

**Comment by maxandersen** [comment]
`source` has no assignable `lines`  (gives undefined method `lines=`)

**Comment by serussell** [comment]
Not being one of the asciidoctor developers, I can't speak to what's idiomatic; however, the authors made Block sources immutable. Rather than trying to modify the source of the block, how about replacing the blocks you're modifying?

You'll probably want to think of a more efficient way of doing this, however:

```ruby
doc.find_by(context: :paragraph).each {|block|
    parent = block.parent
    idx = parent.blocks.index(block)
    new_lines = do_your_awesome_thing( block.source.lines )
    parent.blocks[idx] = Block.new( parent, :paragraph, :source = your_lines )
}
```

**Comment by msc-** [comment]
Change bars and critique markup are 2 different features though they would share a lot of the same tools & code. This issue seems mostly concerned with the latter but the former is also a very important feature for many people.

For change bars you need to run a diff in conjunction with your SCM tool that would output the _entire_ document, not just the changed portions, with markup, while for critiques the markup would be manually added. There would clearly be a lot of overlap in the markup for the 2 cases; they could possibly even be identical.

So 4 things are needed

1. Definition of markup for showing deletions, insertions and changes for both words and blocks.
2. A diff tool to generate a marked up doc from 2 versions of its Asciidoctor source.
3. A tool, which could be built in to Asciidoctor, to convert that markup into markup for the selected output format, e.g, spans, that allows the changes to be styled.
4. CSS additions, and equivalent for other formats, to render the changes using that markup.

**Comment by elextr** [comment]
As you indicate, markup to indicate changes is something different from critique markup.  Better to open a separate issue for it (if there isn't one already).

**Comment by msc-** [comment]
Since there is so much overlap between the 2, I'll wait for @mojavelinux to chime in first. If he wants a separate issue, I'll make one.

**Comment by msc-** [comment]
You can get git diff to output the entire document with something like

```
git diff --word-diff --inter-hunk-context=10000 HEAD~1 HEAD
```

A tool to interpret the change markup in that output would have to recognize that some valid Asciidoctor markup may appear within the change markup, e.g

```
{+=== New Section Title+}
...
{+===+}
```

as well as dealing with the conflicts between the `git diff`markup and Asciidoctor markup.

**Comment by elextr** [comment]
> as well as dealing with the conflicts between the git diffmarkup and Asciidoctor markup.

The tool would have to parse git diff and generate valid Asciidoc, not just copy git diff and hope asciidoc would accept it.

**Comment by msc-** [comment]
> The tool would have to parse git diff and generate valid Asciidoc, not just copy git diff and hope asciidoc would accept it.

Indeed. That is what the section "A tool to interpret the change markup in that output ..." is referring
to.

**Comment by elextr** [comment]
@msc- If it parses git diff and generates asciidoc there is no conflict AFAICT so I interpreted your comment about "dealing with conflict" as it was copying stuff, no problem I think we are agreeing in different terminology :)

**Comment by msc-** [comment]
I simply meant that the tool will have to make sure it does not become confused between the genuine Asciidoctor markup that will be in the `git diff` output and change markup that happens to look like Asciidoctor markup.

**Comment by elextr** [comment]
Ahh, I see where you are coming from, but since git diff is defined so that it can be used to pass patches to any type of text file its guaranteed to be able to be parsed irrespective of the file contents, so shouldn't be a conflict.

**Comment by msc-** [comment]
In my vision, the `git diff` output is the entire document with the `--word-diff` change markup. It is not patches. That said, the fact that + & - are not allowed in Asciidoctor macro or attribute names means it won't be that difficult to avoid confusing {+new text+} with an {attribute_reference}, both of which could be in the `git diff` output.

**Comment by msc-** [comment]
I was thinking of using roles to indicate additions, deletions and changes but I can see no way to set a role for in-line text unless the text is also quoted. So some help is needed from Asciidoctor, such as the critique markup, to make a complete change bar solution.

**Comment by shivakumargn** [comment]
Plain old diff has options to insert custom markups for insertion/deletion.
`diff --new-line-format='+%L' --old-line-format='-%L' --unchanged-line-format=' %L' sample.adoc sample.new.adoc`
There is also the gnu wdiff command which provides `git diff --word-diff` like facility with more customizability for the markup.
`wdiff --start-delete='{-' --end-delete='-}' --start-insert='{+' --end-insert='+}' sample.adoc sample.new.adoc`

`wdiff` can accept unified diff as well (it can accept git diff and sanitize it with custom markup and word diffs required).

Assumption of `git` command to exist for processing or the `git diff` output as the input format may be too restrictive.

**Comment by rohel01** [comment]
> Is change-tracking really the responsibility of a format processor? I respectfully suggest that change-tracking and displaying editorial markup isn't within the proper scope of Asciidoctor. Typical display markup such as [line-through](http://asciidoc.org/userguide.html#X51) is already supported by AsciiDoc, and editorial insertions can be displayed as inline text like `Hi, [she] said.` Editorial markup can also be placed in [comment blocks](http://asciidoc.org/userguide.html#X26) if the critique should be visible in the source document, but omitted from the rendered output.

I agree.  Change tracking is not the responsibility of the text processor and publishing toolchain.  

> Another thing to consider is that editorial markup is rarely appropriate for display purposes. Marking up the markup is useful for communicating about changes that need to be made to the source document, but the practical use cases for _displaying_ the editorial marks in an output document (as opposed to the source document) would seem rare.
> 
> I think there's a great deal of usefulness in allowing for editorial markup in source documents, but very little for such markup in rendered output documents. That's just my $0.02, of course, but I think it's a viewpoint worth considering.

I think precisely the opposite. My team write technical software documentation containing dozens of requirements and interface descriptions. It is customary to publish this kind of documents in at least two flavours: with and without change tracking. Indeed, given the size of these documents, it is critical stakeholders can quickly review changes between release, without reading the complete document each time. So, displaying editorial markup IS useful in that use case.

Please note the displayed editorial markup should not require or rely on Asciidoc syntax or features. Asciidoctor is an implementation detail of the document production process. Stakeholders only care about its result, the edited documents. The fact that my team has used Asciidoctor, LaTeX or Word to publish them should be irrelevant to them. Typically, deletions are marked with a strikethrough, and additions are marked with an underline. Sometimes, different authors' changes are indicated with different colors.

On the other hand, my team will happily collaborate on our document Asciidoc sources with Gitlab or Github. Reviews happen in the scope of Pull Requests, where discussions take place directly inline of the commit diffs. No need for more fancy tooling or markup.

I am looking into ways to generate PDF documents with change highlights. The lack of this feature is currently a showstopper for my employer.

 I initially thought of parsing a git provided unified diff and highlighting the changes myself with AsciidoctorJ. But I am growing more and more convinced that this is not the correct approach: the diff will show a lot of changes that are not meaningful by themselves (table changes, macro parameters...) or irrelevant to the readers. What about directly comparing two versions of rendered document AST?

Does this approach seem reasonable? If so, I can start working on a prototype (I work in Java).

**Comment by maxandersen** [comment]
I think at this point just go ahead and try and see if it sticks. 

See https://github.com/asciidoctor/asciidoctor/issues/2709 of example of higher level diff in markups.

**Comment by yonkeltron** [comment]
Just checking in here since there hasn't been a ton of movement in a very long time. Is there still interest in implementing this as either a proper feature or a plugin of some sort? I'm very interested in using this or something like it.

**Comment by todd-a-jacobs** [comment]
> I think precisely the opposite. While Asciidoctor should not be responsible with tracking the changes, it should help publish document with highlighted changes.

As long as it's understood that change *tracking* is really outside the scope of critic markup, AsciiDcotor, or even Pandoc, I don't really see why it couldn't or shouldn't be supported for rendered output, especially if it's a plugin or doesn't break the current core. While I still think there may be better tools for this (e.g. diff-pdf, or word diffs of the source Markdown or CommonMark documents), I'm not sure that I'm as opposed to it as I once was.

Rendering critic markup in output documents (or even just word diffs for that matter, which might or might not be easier to implement) might be useful in some cases. While any new feature has the potential to create feature bloat, the fact that there's still interest in this feature years after it was originally brought up has caused me to reconsider my stance that it's an anti-feature. So long as people understand the limitations, and so long as it doesn't require refactoring the whole core of AsciiDoc, my non-maintainer position is that it's probably worthy of reconsideration especially if there's an associated pull request by some interested party.

Again, just my $0.02.

**Comment by yonkeltron** [comment]
I understand what you're saying and I agree with your core premise. Now I just need to understand how to go about implementing this sort of feature in AsciiDoctor. Almost feels like it'd be easier to do it as a separate preprocessor before AsciiDoctor even gets involved but that's just my initial hesitation talking.

**Comment by JoelMon** [comment]
Arrived here after searching for a way to critique my doc. Putting my vote for a solution. :+1:

**Comment by todd-a-jacobs** [comment]
@mojavelinux 

> > I would like to lobby for a simple, quickly-typable form of markup in Asciidoctor for line-through
> 
> Given that Markdown has this, I do think this request could fall within the scope of the specification.

While all of critic markup seems more useful now than it did in the past&mdash;at least to me, and I'm sorry I said it was a bad idea at the time&mdash;depending on the flavor some parts may already be available. For example, in iA Write I can do this:

```markdown
This seems more ==~~unnecessary~~ necessary== as it as Markdown and AsciiDoc gain wider adoption.
```

This renders for me in iA Writer s:

<img width="1134" height="134" alt="Image" src="https://github.com/user-attachments/assets/0ba327d7-57d9-4ddc-a856-9314d9e091e2" />

but renders differently (and not usefully) with GFM as:

> This seems more ==~~unnecessary~~ necessary== as it as Markdown and AsciiDoc gain wider adoption.

P.S. For anyone on the "only LLMs write with em-dashes, you probably don't bother with the Oxford comma either. 😢

**Comment by mrshu** [comment]
Sorry to revive this old thread, but this still comes up for us in practice.

We’ve been using CriticMarkup-in-AsciiDoc as an authoring workflow for years: writers keep CriticMarkup markers in an `.adoc` source, then a small preprocessor turns those markers into plain AsciiDoc before Asciidoctor builds HTML and PDF. It’s been working well for iterative edits and reviews.

One concrete place this is used is **[RoboCupJunior Soccer](https://junior.robocup.org/soccer/
) rules** (source-controlled AsciiDoc which is then published as HTML/PDF):
- Rules repo: https://github.com/robocup-junior/soccer-rules
- Source `.adoc`: https://github.com/robocup-junior/soccer-rules/blob/master/rules.adoc
- Published HTML: https://robocup-junior.github.io/soccer-rules/master/rules.html
- Published PDF: https://robocup-junior.github.io/soccer-rules/master/rules.pdf

If it helps anyone following this issue, I recently packaged our preprocessor as a small CLI called `criticmarkup` which perhaps might be useful for some: https://github.com/mrshu/criticmarkup

This approach doesn’t require Asciidoctor itself to parse CriticMarkup -- it just consumes normal AsciiDoc after preprocessing. 

If some more details could be of help, feel free to just ask.

Thanks!
